near-sdk/src/collections/tree_map.rs

```diff
@@ -0,0 +1,1772 @@
+{
+    fn of(id: u64, key: K) -> Self {
+        Self {
+            id,
+            key,
+            lft: None,
+            rgt: None,
+            ht: 1,
+        }
+    }
+}
+
+impl<K, V> Default for TreeMap<K, V>
+    where
+        K: Ord + Copy + BorshSerialize + BorshDeserialize,
+        V: Copy + BorshSerialize + BorshDeserialize,
+{
+    fn default() -> Self {
+        Self::new(next_trie_id())
+    }
+}
+
+
+impl<K, V> TreeMap<K, V>
+    where
+        K: Ord + Copy + BorshSerialize + BorshDeserialize,
+        V: Copy + BorshSerialize + BorshDeserialize,
+{
+    pub fn new(id: Vec<u8>) -> Self {
+        Self {
+            root: 0,
+            val: UnorderedMap::new(append(&id, b'v')),
+            tree: Vector::new(append(&id, b'n')),
+        }
+    }
+
+    pub fn len(&self) -> u64 {
+        self.tree.len() as u64
+    }
+
+    pub fn clear(&mut self) {
+        self.root = 0;
+        self.val.clear();
+        self.tree.clear();
+    }
+
+    fn node(&self, id: u64) -> Option<Node<K>> {
+        self.tree.get(id)
+    }
+
+    fn save(&mut self, node: &Node<K>) {
+        if node.id < self.len() {
+            self.tree.replace(node.id, node);
+        } else {
+            self.tree.push(node);
+        }
+    }
+
+    /// Return height of the tree - number of nodes on the longest path starting from the root node.
+    pub fn height(&self) -> u64 {
+        self.node(self.root).map(|n| n.ht).unwrap_or_default()
+    }
+
+    pub fn contains_key(&self, key: &K) -> bool {
+        self.val.get(key).is_some()
+    }
+
+    pub fn get(&self, key: &K) -> Option<V> {
+        self.val.get(key)
+    }
+
+    pub fn insert(&mut self, key: &K, val: &V) -> Option<V> {
+        if !self.contains_key(&key) {
+            self.root = self.insert_at(self.root, self.len(), &key);
+        }
+        self.val.insert(&key, &val)
+    }
+
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        if self.contains_key(&key) {
+            self.root = self.do_remove(&key);
+            self.val.remove(&key)
+        } else {
+            // no such key, nothing to do
+            None
+        }
+    }
+
+    /// Returns the smallest stored key from the tree
+    pub fn min(&self) -> Option<K> {
+        self.min_at(self.root, self.root).map(|(n, _)| n.key)
+    }
+
+    /// Returns the largest stored key from the tree
+    pub fn max(&self) -> Option<K> {
+        self.max_at(self.root, self.root).map(|(n, _)| n.key)
+    }
+
+    /// Returns the smallest key that is strictly greater than key given as the parameter
+    pub fn above(&self, key: &K) -> Option<K> {
+        self.above_at(self.root, key)
+    }
+
+    /// Returns the largest key that is strictly less than key given as the parameter
+    pub fn below(&self, key: &K) -> Option<K> {
+        self.below_at(self.root, key)
+    }
+
+    /// Returns the smallest key that is greater or equal to key given as the parameter
+    pub fn floor(&self, key: &K) -> Option<K> {
```

Comment:
this is still backwards compared to java TreeMap `floorKey`, `ceilingKey`.
Also one of the doc comments should be `less or equal`

Reply:
[Java](https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html)

```
ceilingKey(K key)
Returns the least key greater than or equal to the given key, or null if there is no such key.

floorKey(K key)
Returns the greatest key less than or equal to the given key, or null if there is no such key.
```

Updating existing to:

```
/// Returns the smallest key that is greater or equal to key given as the parameter
    pub fn ceil_key(&self, key: &K) -> Option<K>

/// Returns the largest key that is less or equal to key given as the parameter
    pub fn floor_key(&self, key: &K) -> Option<K>
```